File: examples/src/test/scala/portals/examples/PingPong.scala
package portals.examples

import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.junit.Assert._
import org.junit.Ignore
import org.junit.Test

import portals.api.builder.ApplicationBuilder
import portals.test.TestUtils
import portals.Systems

@RunWith(classOf[JUnit4])
class PingPongTest:

  @Test
  def testPingPong(): Unit =
    import portals.api.dsl.DSL.*

    case class Ping(i: Int)
    case class Pong(i: Int)

    val tester = new TestUtils.Tester[Pong]()

    val system = Systems.interpreter()

    {
      val pinger = ApplicationBuilder("pinger")

      val sequencer = pinger.sequencers("sequencer").random[Pong]()

      val _ = pinger
        .workflows[Pong, Ping]("workflow")
        .source[Pong](sequencer.stream)
        .map { case Pong(i) => Ping(i) }
        .sink[Ping]()
        .freeze()

      val pingerApp = pinger.build()

      system.launch(pingerApp)
    }

    {
      val ponger = ApplicationBuilder("ponger")

      val extStream = ponger.registry.streams.get[Ping]("/pinger/workflows/workflow/stream")

      val pongerwf = ponger
        .workflows[Ping, Pong]("ponger")
        .source[Ping](extStream)
        .map { case Ping(i) => Pong(i - 1) }
        .filter(_.i > 0)
        // .logger()
        .task(tester.task)
        .sink[Pong]()
        .freeze()

      val extSequencer = ponger.registry.sequencers.get[Pong]("/pinger/sequencers/sequencer")
      val _ = ponger.connections.connect(pongerwf.stream, extSequencer)

      val pongerApp = ponger.build()

      system.launch(pongerApp)
    }

    {
      val builder = ApplicationBuilder("generator")

      val generator = builder.generators.fromList(List(Pong(128)))

      val extSequencer = builder.registry.sequencers.get[Pong]("/pinger/sequencers/sequencer")

      val connection = builder.connections.connect(generator.stream, extSequencer)

      val app = builder.build()

      system.launch(app)
    }

    system.stepUntilComplete()

    system.shutdown()

    assertEquals((127 until 0 by -1).map(Pong(_)).toList, tester.receiveAll())

File: portals-portalsjs/js/src/main/resources/examples/playground/pingPong.js
let builder = PortalsJS.ApplicationBuilder("pingPong")

let portal = builder.portal.portal("pingPongPortal")

let generator = builder.generators.signal(8)

function pingerPonger(ctx, x) {
  if (x > 0) {
    ctx.log.info(x);
    let future = ctx.ask(portal, x - 1);
    ctx.await(future, ctx => {
      let v = future.value(ctx);
      ctx.reply(pingerPonger(ctx, v));
    });
  }
  else {
    ctx.reply(x);
  }
}

let pingPongWorkflow = builder.workflows
  .source(generator.stream)
  .askerreplier(
    portal,
    portal,
    ctx => x => {
      let future = ctx.ask(portal, x);
      ctx.await(future, ctx => {
        ctx.log.info("Done!");
      });
    },
    ctx => x => {
      pingerPonger(ctx, x)
    },
  )
  .logger()
  .sink()
  .freeze()

let pingPong = builder.build()
let system = PortalsJS.System()
system.launch(pingPong)
system.stepUntilComplete()
